spdx-utils/src/test/kotlin/SpdxExpressionTest.kt

```diff
@@ -493,6 +494,57 @@ class SpdxExpressionTest : WordSpec() {
             }
         }
 
+        "applyChoices()" should {
+            "return the correct result if a single choice is applied" {
+                val expression = "a OR b OR c OR d".toSpdx()
+
+                val choices = listOf(LicenseChoice(expression, "a".toSpdx()))
+
+                val result = expression.applyChoices(choices)
+
+                result shouldBe "a".toSpdx()
+            }
+
+            "return the correct result if multiple simple choices are applied" {
+                val expression = "a OR b AND c OR d".toSpdx()
+
+                val choices = listOf(
+                    LicenseChoice("a OR b".toSpdx(), "a".toSpdx()),
+                    LicenseChoice("c OR d".toSpdx(), "c".toSpdx())
+                )
+
+                val result = expression.applyChoices(choices)
+
+                result shouldBe "a AND c".toSpdx()
+            }
+
+            "ignore invalid sub-expressions and return the correct result for valid choices" {
+                val expression = "a OR b OR c OR d".toSpdx()
+
+                val choices = listOf(
+                    LicenseChoice("a OR b".toSpdx(), "b".toSpdx()), // b OR c OR d
+                    LicenseChoice("a OR c".toSpdx(), "a".toSpdx()) // not applied
+                )
+
+                val result = expression.applyChoices(choices)
+
+                result shouldBe "b OR c OR d".toSpdx()
+            }
+
+            "apply the second choice to the effective license after the first choice" {
+                val expression = "a OR b OR c OR d".toSpdx()
+
+                val choices = listOf(
+                    LicenseChoice("a OR b".toSpdx(), "b".toSpdx()), // b OR c OR d
+                    LicenseChoice("b OR c".toSpdx(), "b".toSpdx()) // b OR d
+                )
+
+                val result = expression.applyChoices(choices)
+
+                result shouldBe "b OR d".toSpdx()
+            }
```

Comment:
Could we have a test case where a single choice can be applied multiple times? E.g. we have something like `(a OR b) AND (c OR d) AND (a OR e)`. When applying the choice `a` to the whole expression I assume the result should be `a AND (c OR d) AND a`, and this can probably be simplified to `a AND (c OR d)`. Is this assumption correct and should this be tested as well?

Reply:
Your assumption is incorrect.
`a` is not a valid choice for `(a OR b) AND (c OR d) AND (a OR e)`, because the choice is not just `a OR b` but always need to take the licenses from the `AND` expressions into consideration.

To get your wanted result the user would need to configure something like this:
```
- given: a OR b
  choice: a
- given: a OR e
  choice: a
```

I added a test for this.